### PR TITLE
cds: add a benchmark

### LINKED ISSUE #14005: Create benchmark tests for CDS
*Title*: speed tests for CDS.

*Description*: #10982 introduced a benchmark_test for EDS, I'd like to do the same for CDS, and work on whatever non-linear performance it finds, if any.

### PR DESCRIPTION
**Pull request by pgenera**
Commit Message: Add a benchmark test for Cluster Discovery Service.
Additional Description: This copies eds_speed_test and modifies it to exercise CDS instead, with some small changes in benchmark options.
Risk Level: Low - test only change.
Testing: Test runs & passes in both _benchmark_test & bazel run cds_speed_test variants.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

Fixes #14005

### CODE REVIEW

**Comment by pgenera** [comment]
clang-tidy failure is:

```Value stored to '_' during its initialization is never read```

which is, uh, the point?

**Comment by zuercher** [comment]
We've added `// NOLINT` to [similar lines](https://github.com/envoyproxy/envoy/blob/master/test/common/common/re_speed_test.cc#L33) in other benchmark tests. I think it's fine to do the same here.

File: test/common/upstream/cds_speed_test.cc
Comment (jmarantz) [inline test/common/upstream/cds_speed_test.cc:186]: can you comment a bit on a few choices here:

1. Why call `speed_test.clusterHelper(true, clusters);` twice?
2. Why not construct `CdsSpeedTest` outside the loop?
3. Related: should we put the `Envoy::Logger::Context` into the class to avoid repeating that code for each test?

I'm not pushing back against those decisions I'm just trying to understand them.

**Review by jmarantz — COMMENTED** [review]
This looks great to me though I'm not an expert in the cluster-init stuff so I may have missed something.

Just a few questions which could be resolved by adding comments or small changes.

File: test/common/upstream/cds_speed_test.cc
Comment (yanavlasov) [inline test/common/upstream/cds_speed_test.cc:168]: Wouldn't this also include the time it took to build the response proto in the benchmark?

Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc:168]: I don't believe so; the calls to ```state_.PauseTiming()``` and ```ResumeTiming()``` in clusterHelper should exclude everything but the grpc callback?

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc:180]: 1) that's the difference between the two test cases; I added a comment (hopefully) explaining.
2) I assume the loop here is modifying state in a meaningful way, which is then passed to the newly constructed speed_test. This ```auto _ : state``` is directly from the example in https://github.com/google/benchmark
3) y'know, I don't know what the Logger Context was doing here, and removing it has no ill effect...

This was mostly copy & pasted from eds_speed_test (I don't understand github's copy detection and why it didn't notice this); I've backported the changes to eds_speed_test as well.

File: test/common/upstream/cds_speed_test.cc
Comment (jmarantz) [reply test/common/upstream/cds_speed_test.cc:165]: It would be clearer I think what we were benchmarking if we put the PauseTiming calls at the start of the loop here (line 162) and Resume at the end, and then put Resume/Pause calls in the helper functions surrounding the code we really want to test.

File: test/common/upstream/cds_speed_test.cc
Comment (jmarantz) [reply test/common/upstream/cds_speed_test.cc:180]: RE 1 and 3: cool, all set, thanks.

Re 2: the `auto _ state : state` pattern is fine, but I think state.range(x) won't change within a call, and it would be  safe to construct the CdsSpeedTest object outside the loop. I think that will result in a more focused perf benchmark. WDYT?

Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc:180]: Testing bears out your hypothesis, I see what I missed there.

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc:167]: I think this is what you mean.

**Comment by pgenera** [comment]
I (somewhat unrelatedly) made the floor iteration counts match between cds & eds_speed_test in the last commit; below 100 or so iterations the constant factor dominates.

File: test/common/upstream/cds_speed_test.cc
Comment (jmarantz) [inline test/common/upstream/cds_speed_test.cc:166]: nit: name this "num_clusters" as if it's just "clusters" it sounds like a container. Also hoist this out of the loop, and make it const :)

File: test/common/upstream/cds_speed_test.cc
Comment (jmarantz) [inline test/common/upstream/cds_speed_test.cc:182]: num_clusters, hoist, and const :)

**Review by jmarantz — COMMENTED** [review]
nice! lgtm; just a couple of nits.

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc]: done, here and in eds_speed_test

**Review by jmarantz — APPROVED** [review]
thanks phil!

**Comment by jmarantz** [comment]
Probably it makes sense for some @envoyproxy/senior-maintainers to take a look -- there may be some subtlety in the interaction with the CDS system.

Phil -- did you get some interesting results with the test?

**Comment by pgenera** [comment]
No smoking gun, unfortunately, just things we know: v3 is faster than v2, ignoring unknown fields is faster than validating them. Unlike EDS, there's no obvious n^2 or worse behavior.

```
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
addClusters/0/0/64          0.181 ms        0.182 ms         3855
addClusters/1/0/64          0.691 ms        0.691 ms         1010
addClusters/0/1/64          0.092 ms        0.092 ms         7674
addClusters/1/1/64          0.218 ms        0.219 ms         3146
addClusters/0/0/512          1.36 ms         1.36 ms          514
addClusters/1/0/512          5.55 ms         5.55 ms          125
addClusters/0/1/512         0.626 ms        0.626 ms         1114
addClusters/1/1/512          1.78 ms         1.78 ms          394
addClusters/0/0/4096         11.7 ms         11.7 ms           61
addClusters/1/0/4096         45.9 ms         45.9 ms           15
addClusters/0/1/4096         5.66 ms         5.66 ms          125
addClusters/1/1/4096         15.5 ms         15.5 ms           46
addClusters/0/0/32768        99.6 ms         99.6 ms            6
addClusters/1/0/32768         368 ms          368 ms            2
addClusters/0/1/32768        52.6 ms         52.6 ms           13
addClusters/1/1/32768         127 ms          127 ms            5
addClusters/0/0/100000        300 ms          300 ms            2
addClusters/1/0/100000       1127 ms         1127 ms            1
addClusters/0/1/100000        159 ms          159 ms            4
addClusters/1/1/100000        389 ms          389 ms            2
addClusters_BigO          4934.77 N       4934.10 N    
addClusters_RMS               130 %           130 %    
duplicateUpdate/64          0.187 ms        0.187 ms         3832
duplicateUpdate/512          1.31 ms         1.31 ms          534
duplicateUpdate/4096         11.6 ms         11.6 ms           62
duplicateUpdate/32768         107 ms          107 ms            7
duplicateUpdate/100000        316 ms          316 ms            2
duplicateUpdate_BigO      3172.78 N       3172.41 N    
duplicateUpdate_RMS             2 %             2 %    
```

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [inline test/common/upstream/cds_speed_test.cc:51]: Why is this EDS rather than CDS method?

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [inline test/common/upstream/cds_speed_test.cc:58]: Why point at EDS? You can inline the cluster's load assignment via static endpoints. This will allow this test to focus on specifically the CDS overheads (and be simpler).

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [inline test/common/upstream/cds_speed_test.cc:57]: In terms of `Cluster` configuration, I think there are ultimately many dimensions to consider, but the two I'd start with are number of clusters, number of endpoints per cluster and the choice of load balancing algorithm (everything from WRR to Maglev etc.). This will provide a pretty interesting microbenchmark.

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks for putting this together! Do you have any `perf` profiles and flamegraphs when running some of the longer benchmarks?
/wait

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc]: oversight, fixed. I think this now reflects the CDS API, but I'm not completely confident.

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc]: fixed.

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc]: Right now this exercises the number of clusters for both the v2 & v3 APIs; I propose taking out the v2 part and adding LB algorithm & endpoints per cluster, all for ```STATIC``` clusters. Does that sound like a good next revision?

In the interim I've separated out the v2 & v3 tests to make the output more readable and give the Complexity Computing Magic a better chance at working.

**Comment by pgenera** [comment]
Here's the current output, with warnings about v2 API deprecation elided:

```
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
addV3Clusters/0/64          0.780 ms        0.778 ms          916
addV3Clusters/1/64          0.228 ms        0.226 ms         3083
addV3Clusters/0/512          5.99 ms         5.99 ms          119
addV3Clusters/1/512          1.85 ms         1.85 ms          381
addV3Clusters/0/4096         47.0 ms         47.0 ms           15
addV3Clusters/1/4096         14.0 ms         14.0 ms           51
addV3Clusters/0/32768         375 ms          374 ms            2
addV3Clusters/1/32768         103 ms          103 ms            7
addV3Clusters/0/100000       1161 ms         1161 ms            1
addV3Clusters/1/100000        314 ms          314 ms            2
addV3Clusters_BigO        7369.25 N       7368.19 N    
addV3Clusters_RMS              98 %            98 %    

addV2Clusters/0/64           3.62 ms         3.62 ms          192
addV2Clusters/1/64          0.825 ms        0.824 ms          858
addV2Clusters/0/512          27.6 ms         27.5 ms           25
addV2Clusters/1/512          6.56 ms         6.56 ms          108
addV2Clusters/0/4096          219 ms          219 ms            3
addV2Clusters/1/4096         52.0 ms         52.0 ms           14
addV2Clusters/0/32768        1797 ms         1797 ms            1
addV2Clusters/1/32768         405 ms          405 ms            2
addV2Clusters/1/100000       1227 ms         1227 ms            1
addV2Clusters_BigO       32975.49 N      32970.82 N    
addV2Clusters_RMS             107 %           107 %    

duplicateUpdate/64          0.544 ms        0.542 ms         1346
duplicateUpdate/512          4.13 ms         4.12 ms          167
duplicateUpdate/4096         33.7 ms         33.7 ms           21
duplicateUpdate/32768         244 ms          244 ms            3
duplicateUpdate/100000        775 ms          775 ms            1
duplicateUpdate_BigO      7724.84 N       7723.78 N    
duplicateUpdate_RMS             2 %             2 %    
```

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [reply test/common/upstream/cds_speed_test.cc]: Yes, let's drop v2 as it's on the way out. Agree on the changes proposed.

File: test/common/upstream/eds_speed_test.cc
Comment (htuch) [inline test/common/upstream/eds_speed_test.cc:207]: How come you're changing `eds_speed_test.cc` as well?

**Review by htuch — COMMENTED** [review]
Looks good, waiting for v2 removal.

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [reply test/common/upstream/cds_speed_test.cc]: FWIW, I think we could also explore LB algorithm and endpoints in the EDS test if it was able to fully simulate what is going on in `eds_speed_test.cc`, so maybe we can make CDS test even simpler and exercise some other cluster attributes in the parameter space. For now just dropping v2 sounds good.

File: test/common/upstream/eds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/eds_speed_test.cc:207]: This (should be) a port of jmarantz's style comments from ```cds_speed_test```

File: test/common/upstream/cds_speed_test.cc
Comment (pgenera) [reply test/common/upstream/cds_speed_test.cc]: I've dropped v2; maybe adding a larger set of test parameters to this or cds should go in a followup?

**Comment by pgenera** [comment]
The CI failures appear to be real; I'm taking a look.

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [reply test/common/upstream/cds_speed_test.cc]: Let's do the additions in followup PRs.

File: test/common/upstream/BUILD
Comment (htuch) [inline test/common/upstream/BUILD:174]: Nit: might be worth a quick pass to IWYU and minify any deps that no longer apply. Generally we're not super strict, but this list is pretty long.

File: test/common/upstream/cds_speed_test.cc
Comment (htuch) [inline test/common/upstream/cds_speed_test.cc:103]: Nit: not needed

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo nits. It would be good to see a perf profile to validate this is measuring the right overheads (and not test artifacts) before calling this benchmark complete, but happy to ship and do that after.
/wait

File: test/common/upstream/BUILD
Comment (pgenera) [reply test/common/upstream/BUILD:174]: I didn't manage to run IWYU successfully, but I did a pass and removed the obviously vestigial entries.

**Review by htuch — DISMISSED** [review]
LGTM, thanks. I think the CDS/EDS tests could ultimately be refactoring to make a framework for more general xDS benchmarks, but we can live by Rule of Three in this PR.

**Comment by htuch** [comment]
@pgenera the ASAN failure looks legit.

**Comment by pgenera** [comment]
I've managed to get a [flame graph](https://gist.github.com/pgenera/dd207092d1c14bb73ac3c3ecbd47782e) (SVG cannot be attached here, so have a gist) out, via the following:
```
perf record -g ~/git/envoy/bazel-bin/test/common/upstream/cds_speed_test -- --benchmark_filter="addClusters/1/100000" --benchmark_repitions=10
perf script | ~/FlameGraph/stackcollapse-perf.pl  > out.perf
cat out.perf ~/FlameGraph/flamegraph.pl > cds.svg
```

I'll take a look at the asan failure shortly.

**Comment by htuch** [comment]
Thanks @pgenera. Almost all the time is in `buildStaticCluster`, which is not in the measured benchmark portion but still exits. I wonder if we can build all clusters once and then do a ton of updates to avoid having this be excessive. This will allow for actionable use of the benchmark to dive into the CDS overheads via profiles.

**Comment by pgenera** [comment]
I was going to explore filtering the `perf` output, but yeah, that's annoying (it also explains why this test takes so long to run despite the timings being quite small).

Separately I wonder if these cluster creation helpers should create protos directly instead of parsing YAML.

**Comment by jmarantz** [comment]
Yeah the flame-graph you shared is dominated by the protobuf parsing library which is not consistent with what we see from captures from live servers.

Probably the actual protocol marshall/unmarshall is much faster than yaml parsing.

**Comment by htuch** [comment]
> Separately I wonder if these cluster creation helpers should create protos directly instead of parsing YAML.

This would be optimal. I'd just port the `buildStaticCluster` logic to your benchmark, it's not doing _that_ much. That utility is intended for integration testing (and clarity via YAML) not for performance. I'd probably recommend against using it anyway, since performance and testing have divergent requirements.

**Comment by jmarantz** [comment]
That sounds good. Would it be sufficient though to just turn off the profile collection during the yaml parsing,. assuming that's possible?

**Comment by pgenera** [comment]
I ported buildStaticCluster into the benchmark and got a more interesting [flame graph](https://gist.githubusercontent.com/pgenera/a2f5de450c37cc004e9f8126a675ea00/raw/80d2a3732d6cb5fdd3b3ccecd786f263f38e044e/cds.svg) out; I haven't dug deep or done any filtering yet however.

I agree that we're on the cusp of wanting to build something generic for xDS benchmarking.

**Comment by pgenera** [comment]
Here's the most recent [flame graph](https://gist.github.com/pgenera/f34b00af7036d898d139be30316821c7), now using a more concrete stats implementation; however I still don't see the stats overhead I'd expect.

I don't have any leads on the asan failure, other than having reproduced it:
```
$ bazel build -c dbg --config=clang-asan //test/common/upstream:eds_speed_test_benchmark_test
...
ERROR: /usr/local/google/home/pgenera/git/envoy/test/common/upstream/BUILD:114:26: Linking of rule '//test/common/upstream:eds_speed_test' failed (Exit 1) clang failed: error executing command /usr/local/google/home/pgenera/.clang9/bin/clang @bazel-out/k8-dbg/bin/test/common/upstream/eds_speed_test-2.params

Use --sandbox_debug to see verbose messages from the sandbox
ld.lld: error: undefined symbol: __muloti4
>>> referenced by int128_have_intrinsic.inc:251 (external/com_google_absl/absl/numeric/int128_have_intrinsic.inc:251)
>>>               bazel-out/k8-dbg/bin/external/com_google_absl/absl/strings/_objs/strings/numbers.pic.o:(absl::operator*(absl::int128, absl::int128))
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
Target //test/common/upstream:eds_speed_test_benchmark_test failed to build
```

**Comment by pgenera** [comment]
Having now found #13973, it builds with `--config=asan` and `-fsanitize=undefined` removed from the project bazelrc. Obviously that's not the right solution, but its at least informative.

**Comment by jmarantz** [comment]
Quick sanity check: did you use "-c opt" for the build you used to collect the flame graph?

**Comment by pgenera** [comment]
Can confirm:
```
$ history | grep -B 4 "perf record"
 1920  [2020-12-11 20:06:26 +0000] bazel build -c opt
//test/common/upstream:cds_speed_test
 1921  [2020-12-11 20:08:55 +0000] perf record -g
./git/envoy/bazel-bin/test/common/upstream/cds_speed_test --
--benchmark_filter="addClusters/1/100000" --benchmark_repitions=100
 1922  [2020-12-11 20:08:57 +0000] cd
 1923  [2020-12-11 20:08:59 +0000] perf record -g
./git/envoy/bazel-bin/test/common/upstream/cds_speed_test --
--benchmark_filter="addClusters/1/100000" --benchmark_repitions=100
```

**Comment by htuch** [comment]
I don't think it's an issue of stats or not, we're not spending any significant time in https://github.com/envoyproxy/envoy/blob/046b0e315e04f645aa0d62119a056bc560497ed7/source/common/upstream/cds_api_impl.cc#L84. This is what dominates in the workload flamegraphs that we want to model. I'm suspecting some artifact of the benchmark, e.g. mocks or maybe the actual config proto, is to blame. I suggest doing a single iteration run with `-l trace` logging and try figure out what is going on in terms of code execution, i.e. is a single cluster being built as expected.

**Comment by jmarantz** [comment]
I'm also wondering if it might be useful to start swapping out the mocks
for real implementations.

One alternative approach is to use the MainCommon class, which is a
linkable way to set up all the production classes; no mocks.

**Comment by pgenera** [comment]
Sounds good, I'll start tearing the mocks out.

**Pull request closed without merging**
